### PR TITLE
SDSTOR-10892 : concurrent mem btree UT

### PR DESCRIPTION
**Pull request by shosseinimotlagh**

### CODE REVIEW

File: src/include/homestore/btree/btree.hpp [hunk 1]
```diff
@@ -64,13 +64,22 @@ class Btree {
...
     static BtreeThreadVariables* bt_thread_vars() {
-        static thread_local BtreeThreadVariables* s_ptr{nullptr};
-        if (s_ptr == nullptr) {
-            static thread_local BtreeThreadVariables inst;
-            s_ptr = &inst;
-        }
-        return s_ptr;
+        //static boost::fibers::local_fiber <BtreeThreadVariables> *s_ptr{nullptr};
+        auto this_id( boost::this_fiber::get_id() );
+        static thread_local std::map<fibers::fiber::id,  std::unique_ptr<BtreeThreadVariables>> fiber_map;
```
Comment (sanebay) [inline src/include/homestore/btree/btree.hpp:78]: Does this assume that fibers are not moved across threads. In that case fiber1 will have entry in multiple threads.

File: src/tests/test_mem_btree.cpp [hunk 2]
```diff
@@ -507,6 +514,352 @@ TYPED_TEST(BtreeTest, RandomRemoveRange) {
...
+        m_test_count = m_fibers.size();
+        iomanager.run_on_wait(m_fibers[0], [preload_size, this]() { preload(0, preload_size); });
+        LOGINFO("PRELOAD DONE");
+        RunInParallel();
+    }
+
+private:
+    void set_operations(std::vector< std::string > op_list) { this->m_op_list = op_list; }
+
+    void run_one_iteration(std::string operation_name) {
```
Comment (sanebay) [inline src/tests/test_mem_btree.cpp:572]: nit: cons string&

File: src/tests/test_mem_btree.cpp [hunk 3]
```diff
@@ -507,6 +514,352 @@ TYPED_TEST(BtreeTest, RandomRemoveRange) {
...
+        uint32_t nkeys = s_rand_range_generator(g_re);
+        int key = -1;
+
+        if (replace) {
+            key = m_range_scheduler.pick_random_existing_keys(nkeys, m_max_range_input);
+        } else {
+            key = m_range_scheduler.pick_random_non_existing_keys(nkeys, m_max_range_input);
+        }
+
+        if (key == -1) { return; }
```
[the remark this answers is not in the document]
Comment (shosseinimotlagh) [reply src/tests/test_mem_btree.cpp:698]: no assert is needed. It is possible that in this range, there is no available free keys (or after multiple random attempts) . So lets return it.

File: src/tests/test_mem_btree.cpp [hunk 4]
```diff
@@ -507,6 +514,352 @@ TYPED_TEST(BtreeTest, RandomRemoveRange) {
...
+            std::unique_lock< mutex > lk(map_lock);
+            std::string x = "";
+            for (auto cur = key; cur < key + nkeys; cur++) {
+                if (data.find(cur) != data.end()) x += fmt::format("[{}]={}\n", cur, data[cur]);
+            }
+            return x;
+        }
+
+    private:
+        bool non_of_them_existed(uint32_t key, uint32_t nkeys) {
```
[the remark this answers is not in the document]
Comment (shosseinimotlagh) [reply src/tests/test_mem_btree.cpp]: same as `nit: std::all_of`

File: src/include/homestore/btree/btree.hpp [hunk 5]
```diff
@@ -37,6 +37,12 @@ typedef std::function< bool(const BtreeKey&, const BtreeKey&, const BtreeValue&,
 
 using BtreeNodePtr = boost::intrusive_ptr< BtreeNode >;
 
+struct BtreeVisualizeVariables {
```
Comment (sanebay) [inline src/include/homestore/btree/btree.hpp:40]: Lets keep this in a separate CL.

File: src/include/homestore/btree/btree.hpp [hunk 6]
```diff
@@ -64,13 +70,23 @@ class Btree {
 
     // This workaround of BtreeThreadVariables is needed instead of directly declaring statics
     // to overcome the gcc bug, pointer here: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66944
-    static BtreeThreadVariables* bt_thread_vars() {
-        static thread_local BtreeThreadVariables* s_ptr{nullptr};
-        if (s_ptr == nullptr) {
-            static thread_local BtreeThreadVariables inst;
-            s_ptr = &inst;
+#if 0
```
Comment (sanebay) [inline src/include/homestore/btree/btree.hpp:73]: remove this ?

File: src/include/homestore/btree/btree.ipp [hunk 7]
```diff
@@ -327,11 +327,12 @@ void Btree< K, V >::print_tree(const std::string& file) const {
     to_string(m_root_node_info.bnode_id(), buf);
     m_btree_lock.unlock_shared();
 
-    BT_LOG(INFO, "Pre order traversal of tree:\n<{}>", buf);
```
Comment (sanebay) [inline src/include/homestore/btree/btree.ipp:330]: we can log the btree to 2 places stdout/BT_LOG and file. file is mostly just for testing.

File: src/include/homestore/btree/btree.ipp [hunk 8]
```diff
@@ -345,6 +346,46 @@ void Btree< K, V >::print_tree_keys() const {
     BT_LOG(INFO, "Pre order traversal of tree:\n<{}>", buf);
 }
 
+template < typename K, typename V >
```
Comment (sanebay) [inline src/include/homestore/btree/btree.ipp:349]: Please create a separate CL for the visualization

File: src/tests/test_index_btree.cpp [hunk 9]
```diff
@@ -46,7 +46,7 @@ chunk_selector_type_t test_common::HSTestHelper::s_ds_chunk_sel_type;
 
 SISL_OPTION_GROUP(test_index_btree,
                   (num_iters, "", "num_iters", "number of iterations for rand ops",
-                   ::cxxopts::value< uint32_t >()->default_value("65536"), "number"),
+                   ::cxxopts::value< uint32_t >()->default_value("1500"), "number"),
```
Comment (sanebay) [inline src/tests/test_index_btree.cpp:49]: with 4k , 65k entries should work right ?

**Review by sanebay — APPROVED** [review]

File: src/tests/test_index_btree.cpp (hunk 9 above)
Comment (shosseinimotlagh) [reply src/tests/test_index_btree.cpp:49]: correct. considering 4 btree node variations, this is for building purpose to make it fast.

**Review by sanebay — APPROVED** [review]
